viam-cartographer/src/slam_service/slam_service.cc

```diff
@@ -186,6 +168,52 @@ ::grpc::Status SLAMServiceImpl::GetInternalState(
     return grpc::Status::OK;
 }
 
+int SLAMServiceImpl::GetPointCloudMapC(
+    viam_carto_get_point_cloud_map_response *response) {
+    std::string pointcloud_map;
+    // Write or grab the latest pointcloud map in form of a string
+    try {
+        std::shared_lock optimization_lock{optimization_shared_mutex,
+                                           std::defer_lock};
+        if (action_mode != ActionMode::LOCALIZING &&
+            optimization_lock.try_lock()) {
+            // We are able to lock the optimization_shared_mutex, which means
+            // that the optimization is not ongoing and we can grab the newest
+            // map
+            GetLatestSampledPointCloudMapString(pointcloud_map);
+            std::lock_guard<std::mutex> lk(viam_response_mutex);
+            latest_pointcloud_map = pointcloud_map;
+        } else {
+            // Either we are in localization mode or we couldn't lock the mutex
+            // which means the optimization process locked it and we need to use
+            // the backed up latest map
+            if (action_mode == ActionMode::LOCALIZING) {
+                LOG(INFO)
+                    << "In localization mode, using cached pointcloud map";
+            } else {
+                LOG(INFO)
+                    << "Optimization is occuring, using cached pointcloud map";
+            }
+            std::lock_guard<std::mutex> lk(viam_response_mutex);
+            pointcloud_map = latest_pointcloud_map;
+        }
+    } catch (std::exception &e) {
+        LOG(ERROR) << "Stopping Cartographer: error encoding pointcloud: "
+                   << e.what();
+        std::terminate();
+    }
+
+    if (pointcloud_map.empty()) {
+        LOG(ERROR) << "map pointcloud does not have points yet";
+        return 1;
+    }
+
+    response->point_cloud_pcd =
+        blk2bstr((const char *)pointcloud_map.c_str(), pointcloud_map.length());
```

Comment:
Is this cast really necessary? You don't seem to need to do it here: https://github.com/viamrobotics/viam-cartographer/pull/121/files#diff-92ff8a5cd3a57430d78c6ba3b61a91d1810399b77613eece3c55841868d2822eR214